Add limitations guide (#434)
meilisearch/documentation: guides/advanced_guides/known_limitations.md

# Known limitations

MeiliSearch has known limitations, some of which may be addressed in future versions

## Design limitations

### Database size

Default database maximum size is 100GiB.
MeiliSearch uses LMDB under the hood as the storage engine. On launch, LMDB need to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory, which is different from the real memory. You can read more [about storage](/resources/about_storage.md).
You can modify this default limit using the options `--max-mdb-size` & `--max-udb-size` as described in the [configuration guide](/guides/advanced_guides/configuration.md#max-mdb-size).

### Number of indexes

> You can create up to 200 indexes in MeiliSearch. When using LMDB, MeiliSearch have to specify a number of databases that must stay low for performance reasons.

Reviewer: The total number of LMDB databases as been hard set to 3000 and an index takes something like 11 databases (IIRC) so `3000 / 11 ~= 272`.

> You can create up to 200 indexes in MeiliSearch. When using LMDB, MeiliSearch have to specify a number of databases that must stay low for performance reasons.

Author: Can one of you confirm that we are using 11 databases in LMDB for one index?

> You can create up to 200 indexes in MeiliSearch. When using LMDB, MeiliSearch have to specify a number of databases that must stay low for performance reasons.

Reviewer: [It does seems like we use 11 databases](https://github.com/meilisearch/MeiliSearch/blob/cb267b68edc2481a1b4985cbd2277cb6cab3b6c9/meilisearch-core/src/store/mod.rs#L210-L221), indeed.

# Known limitations

> MeiliSearch has known limitations, some of which may be addressed in future versions

Reviewer 2: ```suggestion
MeiliSearch has known limitations, some of which may be addressed in future versions.
```

## Design limitations

### Database size

> MeiliSearch uses two database: one for storage and the second one for updates.

Reviewer 2: ```suggestion
MeiliSearch uses two databases: the first one for storage and the second one for updates.
```

> Default database maximum size is 100GiB.

Reviewer 2: ```suggestion
The maximum size of the database is by default 100GiB.
```

> MeiliSearch uses LMDB under the hood as the storage engine. On launch, LMDB need to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory, which is different from the real memory. You can read more [about storage](/resources/about_storage.md).

Reviewer 2: ```suggestion
MeiliSearch uses LMDB under the hood as the storage engine. On launch, LMDB needs to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory, which is different from real memory. You can read more [about storage](/resources/about_storage.md).
```

> MeiliSearch uses LMDB under the hood as the storage engine. On launch, LMDB need to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory, which is different from the real memory. You can read more [about storage](/resources/about_storage.md).

Author: 👍

You can modify this default limit using the options `--max-mdb-size` & `--max-udb-size` as described in the [configuration guide](/guides/advanced_guides/configuration.md#max-mdb-size).

### Number of indexes

> You can create up to 200 indexes in MeiliSearch. When using LMDB, MeiliSearch have to specify a number of databases that must stay low for performance reasons.

Reviewer 3: ```suggestion
You can create up to 200 indexes in MeiliSearch. When using LMDB, MeiliSearch has to specify a number of databases that must stay low for performance reasons.
```

Reviewer 3: >  When using LMDB, MeiliSearch has to specify a number of databases that must stay low for performance reasons. 

I do not understand this sentence :s

Reviewer 2: Me neither. Does it mean to impose a limit of X GiB on the size of these databases or on the number of indexes?

Author: I re-wrote the whole paragraph

### Maximum words per attribute

> A maximum of 1000 words par attribute can be indexed. This limit is enforced for relevancy reasons. The more words there are in a given attribute, the less relevant the search queries will be.

Reviewer 3: What do you mean by attribute ?
<img width="430" alt="Screenshot 2020-08-18 at 11 04 15" src="https://user-images.githubusercontent.com/33010418/90493360-9315ec00-e142-11ea-9ff3-836a0319b20b.png">

Do you mean the `key` or the value of a field?

Author: For clarity I think `1000 words per attribute` is more understandable. Or `1000 word per key`? I won't write `1000 words for a field's value` 😅
What do you think?

Reviewer 3: I would suggest we change this part with the following suggestion

```
A field can contain a maximum of 1000 words. If you put more than 1000 words in one attribute, only the first 1000 will be indexed.

This limit is enforced for relevancy reasons as the more words there are in a given attribute, the less relevant the search queries will be.
```

Also, is this only for relevancy reasons or also for perfomance reasons ?

Reviewer 3: I understand your point, I  would like @matboivin opinion on it. But, could you answer on the newer suggestion?

If you put more than 1000 words in one attribute, only the first 1000 will be indexed.

## Other limitations

### Payload size

> The default limit for the payload size is around 10MB. [This limit can be modified](/guides/advanced_guides/configuration.md#options).

Reviewer 3: ```suggestion
The default limit for the payload size is around 10MB. [This limit can be modified](/guides/main_concepts/documents.html#upload).

When trying to upload a body that is more than 10MB it will raise an error unless you have raised the the limit.

\```bash
$ ./meilisearch --http-payload-size-limit=100000000
\```
> The payload limit is now +-100MB instead of 10MB
  
```

Backslashes should be removed but where in conflict with githubs markdown

Author: I am not sure this is the place we want to explain MeiliSearch's behavior

# Known limitations

MeiliSearch has known limitations, some of which may be addressed in future versions.

Some of these limitations have been made by MeiliSearch developers for relevacy and/or performance while others limitations are enforced by LMDB, the key-value store used by MeiliSearch.

## Design limitations

### Database size

MeiliSearch uses two databases: the first one for storage and the second one for updates.
On launch LMDB needs to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory.

The maxime database size is by default __100GiB__ for each databases. This size can me modified using the options `--max-mdb-size` & `--max-udb-size` as described in the [configuration guide](/guides/advanced_guides/configuration.md#max-mdb-size).

### Number of indexes

You can create __up to 200 indexes__ in MeiliSearch. This limit has been hard set for performance reasons.

### Maximum words per attribute

> A maximum of __1000 words par attribute__ can be indexed. This limit is enforced for relevancy reasons. The more words there are in a given attribute, the less relevant the search queries will be.
> If you put more than 1000 words in one attribute, only the first 1000 will be indexed.

Reviewer 3: ```suggestion
A maximum of __1000 words par attribute__ can be indexed. This means that if an attribute contains more than 1000 words, only the first 1000 words will be indexed and the rest will be silently ignored.

This limit is enforced for relevancy reasons. The more words there are in a given attribute, the less relevant the search queries will be.

```

# Known limitations

MeiliSearch has known limitations, some of which may be addressed in future versions.

Some of these limitations have been made by MeiliSearch developers for relevacy and/or performance while others limitations are enforced by LMDB, the key-value store used by MeiliSearch.

## Design limitations

### Database size

MeiliSearch uses two databases: the first one for storage and the second one for updates.
On launch LMDB needs to know the size that it can allocate on disk. This space will be reserved on disk for LMDB, thus MeiliSearch. This space will also be allocated as virtual memory.

The maxime database size is by default __100GiB__ for each databases. This size can me modified using the options `--max-mdb-size` & `--max-udb-size` as described in the [configuration guide](/guides/advanced_guides/configuration.md#max-mdb-size).

### Number of indexes

You can create __up to 200 indexes__ in MeiliSearch. This limit has been hard set for performance reasons.

> ### Maximum words per attribute

Reviewer 2: (change if you agree if the previous suggestion)
```suggestion
### Maximum values per attribute
```

Reviewer 3: Per my answer on the previous suggestion I disagree with this suggestion

> A maximum of __1000 words par attribute__ can be indexed. This means that if an attribute contains more than 1000 words, only the first 1000 words will be indexed and the rest will be silently ignored.

Reviewer 2: I would keep key/value terms.
```suggestion
A maximum of __1000 values attached to one attribute__ is set. This means that if an attribute contains more than 1000 values, only the first 1000 values will be indexed and the rest will be silently ignored.
```

Reviewer 3: I think when thomas uses the word "indexes", he does not say "you can only have 1000 words per attribute" but "MeiliSearch can only **index** 1000 words per attributes". 
After which we explain what this imply. 

Also, there is a difference between values and word. A word is a group of chars inside a string separated by a white space: 
`wordOne wordtwo ..`

Whereas a value is the whole data linked to an attributes: 

```json
{
 "attribute": value
}
```

So, 1000 values for one attribute is technically not possible because an attribute can only have one value.

Reviewer 2: Oh I see!

Reviewer 2: I updated my suggestion. However, a longer explanation would be welcome.

Reviewer 3: Do you mean I should explain the problem better or this part in the article should be explained more in depth?

I think all those limitation should be added to the relative guides. As for example this information should be added to the `documents` guide. Where in the context of the guide it makes more sense. 

Also, we could add a sort of definition in the glossary for the word "indexation".
What do you think of that?

```
In meilisearch, documents are __indexed__. Indexing is the process of saving data in a database in a prefered way to fit your purpose. In MeilISearch's case, the purpose is to optimise word-based search. Because of that, in MeiliSearch the indexation is centered arround words. 
```

Reviewer 2: I agree about adding limitations to each relative guide. And yes it was about clarifying "indexation" term so that would be a good idea to add it to the glossary.

> This limit is enforced for relevancy reasons. The more words there are in a given attribute, the less relevant the search queries will be.

Reviewer 2: (change if you agree if the previous suggestion)
```suggestion
This limit is enforced for relevancy reasons. The more values are linked to a given attribute, the less relevant the search queries will be.
```

Reviewer 3: Per my answer on the previous suggestion I disagree with this suggestion
